ambry-api/src/main/java/com/github/ambry/config/CompressionConfig.java

```diff
@@ -0,0 +1,289 @@
+ * The configuration applies to compression only.  Decompression is automatic and has no configuration.
+ */
+public class CompressionConfig {
+
+  // Boolean whether compression is enabled in PUT operation.
+  private static final String CONFIG_COMPRESSION_ENABLED = "router.compression.enabled";
+  private static final boolean DEFAULT_COMPRESSION_ENABLED = true;
+
+  // Whether to skip compression if content-encoding present.
+  private static final String CONFIG_SKIP_CONTENT_ENCODING = "router.compression.skip.content.encoding";
+  private static final boolean DEFAULT_SKIP_CONTENT_ENCODING = true;
+
+  // A comma-separated list of compressible fully content-types, like "text/javascript, text/log; Charset=UTF8"
+  private static final String CONFIG_CONTENT_TYPES_COMPRESSIBLE = "router.compression.content.types.compressible";
+  private static final String DEFAULT_CONTENT_TYPES_COMPRESSIBLE = null;
+
+  // A comma-separated list of incompressible fully content-types, like "text/javascript, text/log; Charset=UTF8"
+  private static final String CONFIG_CONTENT_TYPES_INCOMPRESSIBLE = "router.compression.content.types.incompressible";
+  private static final String DEFAULT_CONTENT_TYPES_INCOMPRESSIBLE = null;
+
+  // A comma-separated list of compressible content-prefixes, such as "text" for text/*
+  private static final String CONFIG_CONTENT_PREFIXES_COMPRESSIBLE = "router.compression.content.prefixes.compressible";
+  private static final String DEFAULT_CONTENT_PREFIXES_COMPRESSIBLE = "text";
+
+  // A comma-separated list of incompressible content-prefixes, such as "image" for image/*.
+  private static final String CONFIG_CONTENT_PREFIXES_INCOMPRESSIBLE = "router.compression.content.prefixes.incompressible";
+  private static final String DEFAULT_CONTENT_PREFIXES_INCOMPRESSIBLE = "image";
+
+  // Boolean whether to compress other content-types not specified in content-types and content-prefixes.
+  // It applies to PUT operation only.
+  private static final String CONFIG_COMPRESS_OTHER_CONTENT_TYPE = "router.compression.other.content.types";
+  private static final boolean DEFAULT_COMPRESS_OTHER_CONTENT_TYPE = false;
+
+  // The minimal source/chunk size in bytes in order to qualify for compression.
+  private static final String CONFIG_MINIMAL_DATA_SIZE_IN_BYTES = "router.compression.minimal.content.size";
+  private static final int DEFAULT_MINIMAL_DATA_SIZE_IN_BYTES = 1024;
+
+  // The minimal compression ratio (uncompressedSize/compressSize) in order to keep compressed content.
+  // The higher the compression ratio the better with 1 means the same size.
+  private static final String CONFIG_MINIMAL_COMPRESS_RATIO = "router.compression.minimal.ratio";
+  private static final float DEFAULT_MINIMAL_COMPRESS_RATIO = 1.1f;
+
+  // The name of the algorithm to use.  This name must be pre-registered in CompressionMap class.
+  private static final String CONFIG_ALGORITHM_NAME = "router.compression.algorithm.name";
```

Comment:
So does this mean that everytime we try to compress data, we check if the compression ratio >= "router.compression.minimal.ratio"? If not, we save the data uncompressed?

Reply:
Yes, if the compression has too little value, don't waste the effort of decompression.  So if the compression result is insignificant (incompressible content), then don't use compression.  Just store the blob as raw binary.

For a simple blob, the compression ratio determines whether to keep the compressed data or discard compression and store raw.  For composite blobs, the metadata chunk isn't compressed.  The first data chunk is compressed and check its compression ratio.  If it's good, the compressed data is kept and all other chunks will be compressed.  If the first chunk is incompressible, compressed data will be discarded and all other chunks will not be compressed.   The reason for this behavior is due to there is only 1 blob properties per blob, not per chunk.